### ISSUE TITLE
com.atlassian.confluence.ext.code.render.InvalidValueException

### ISSUE THREAD

**Issue by stefanitsky** [docToolchain/docToolchain#1190]
**Describe the bug**
Confluence code macros is not working.

**To Reproduce**
1. Add any code block with ascii markup;
2. Publish to Confluence;
3. See error.

**Expected behavior**

Code block works fine as it should be.

**Screenshots**
<img width="727" alt="image" src="https://github.com/docToolchain/docToolchain/assets/22547214/91edb8e6-da9f-44b5-b5d5-3cd573f29133">

```
[source]
----
{
    "status": false,
    "data": {},
    "message": "Could not find user account"
}
----
```

**Configuration**
 - docToolchain version: ng
 - OS: OSX + docker (custom image)
 - Confluence version: 7.19.2

**Additional context**

I tried with the language code, without, nothing helps.

Also, i tried to patch [asciidoc2confluence.groovy](https://github.com/docToolchain/docToolchain/blob/ng/scripts/asciidoc2confluence.groovy#L459) by taking an older version of code block template, no changes.

**Comment by arutscheidt** [comment]
We use the Cloud verson and in legacy or new editor mode we have no language support nor linebreaks.
Linebreaks are a luck issue. sometimes it works, but not reproduceable.

All code goes to one line.
![20230526_125524](https://github.com/docToolchain/docToolchain/assets/60868966/c6fb6481-9c4a-43fb-af69-ed0a00fd7c77)

Language support is limited as described in the documentation of publishToConfluence. However, both editor modes support many languages, even groovy.

**Comment by PacoVK** [comment]
@arutscheidt  could you please verify this issue still persists? i could not reproduce it using confluence cloud...

**Comment by PacoVK** [comment]
could also not verify using the old version of the editor using the code of #1189

**Comment by stefanitsky** [comment]
> could also not verify using the old version of the editor using the code of #1189

I got another error:
```
* Where:
Script '/home/dtcuser/docToolchain/scripts/publishToConfluence.gradle' line: 30

* What went wrong:
Execution failed for task ':publishToConfluence'.
> No signature of method: asciidoc2confluence.realTitle() is applicable for argument types: (String) values: [6.4. AD GUI]
  Possible solutions: evaluate(java.lang.String)
```

Error line:
```
=== AD GUI
```

**Comment by stefanitsky** [comment]
hm, works, but i still cannot test #1189 on cda10c3355b34bf5388275449229d22193482f75 commit (last working version before last updates 3 days ago)

**Comment by davbader** [comment]
I can provide you with some insight on the last error message. However, I am unsure if this is related to the actual issue topic. If that is not the case, we can open a new Issue.

The error `No signature of method: asciidoc2confluence.realTitle() is applicable for argument types: (String) values` is caused by function `rewriteInternalLinks()`. 

In 0a49348096cf2a465583b8cce2f14a33107a66a9 it changed from being a closure to a function, which broke the code. Internally, `rewriteInternalLinks` tries to call `realTitle`, which is a closure and outside of the functions scope.

The issue is only present if you have anchor links within your document.

Sadly, I was not able to provide a quick fix, because I do not really know groovy. All solutions always resulted in having to pass the `confluencePagePrefix` and `confluencePageSuffix` variables to lots of chained function calls.

For me it seems like `asciidoc2confluence` could use some refactoring, maybe convert it to a class to be able to access member variables such as  `confluencePagePrefix` and `confluencePageSuffix`.

Maybe @PacoVK can provide insight why he changed a lot of closures to functions?

**Comment by PacoVK** [comment]
@davidschottel  i am a bit confused about your analysis, although i am thankful for your investigation! 
Main reason i changed to functions was to ease a bit testing of the script logic. 

> rewriteInternalLinks tries to call realTitle, which is a closure 

`realTitle` is a function as well as `rewriteInternalLinks`, they should be in the same scope. The error hints that the signature seems to be broken for the input. While type inference says object, we may force to String. 

> For me it seems like asciidoc2confluence could use some refactoring, maybe convert it to a class

Yes, agree, that's basically the direction #1205 goes into. I just recognized the upgrade needs significant changes, and i'll probably try to go class-based route here. 

@stefanitsky i am also confused about your comment :D Does it work now? Any other issue you are facing here? 

Thanks for helping out and reporting the issues that you are facing! Really appreciated

**Comment by davbader** [comment]
I am sorry, if I missunderstand.

I think this is a closure, and `rewriteInternalLinks` will not have access to it.

https://github.com/docToolchain/docToolchain/blob/dc2473f3399c22ec6968a6bad40861e51e0b55b2/scripts/asciidoc2confluence.groovy#L221-L223

**Comment by PacoVK** [comment]
@davidschottel oh wow now it makes sense. Sorry i looked at the wrong source, and ofc you are totally right. I need to check and probably rewrite this to a function, as i [already did](https://github.com/docToolchain/docToolchain/pull/1226/files#diff-2276d2170967c2f4b9b28e18ccc9b9bcaa746ff9e6a915d20666007f780499b7R163) in #1205  
The trick here is to expose variables like `confluencePagePrefix` into the contexts via [`@Field`](https://docs.groovy-lang.org/3.0.7/html/gapi/groovy/transform/Field.html) this omits the necessity to pass everything down along the closure chains :)

**Comment by PacoVK** [comment]
Since #1205 has been merged docToolchain 3.1.1 has the fix 👍 thanks a lot for helping ♥️
